homeassistant/util/template.py

```diff
@@ -0,0 +1,86 @@
+"""
+homeassistant.util.template
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Template utility methods for rendering strings with HA data.
+"""
+# pylint: disable=too-few-public-methods
+import json
+from jinja2.sandbox import ImmutableSandboxedEnvironment
+
+
+def render_with_possible_json_value(hass, template, value):
+    """ Renders template with value exposed.
+        If valid JSON will expose value_json too. """
+    variables = {
+        'value': value
+    }
+    try:
+        variables['value_json'] = json.loads(value)
+    except ValueError:
+        pass
+
+    return render(hass, template, variables)
+
+
+def render(hass, template, variables=None, **kwargs):
+    """ Render given template. """
+    if variables is not None:
+        kwargs.update(variables)
+
+    return ENV.from_string(template, {
```

Comment:
Hmm, the second template should not raise an exception. Check the test with
invalid json for example.

Yes, you're right we should wrap the exceptions.

On Fri, Dec 11, 2015, 12:33 Philip Lundrigan <notifications@github.com>
wrote:

> In homeassistant/util/template.py
> <https://github.com/balloob/home-assistant/pull/720#discussion_r47401827>:
>
> > +        'value': value
> > +    }
> > +    try:
> > +        variables['value_json'] = json.loads(value)
> > +    except ValueError:
> > +        pass
> > +
> > +    return render(hass, template, variables)
> > +
> > +
> > +def render(hass, template, variables=None, **kwargs):
> > +    """ Render given template. """
> > +    if variables is not None:
> > +        kwargs.update(variables)
> > +
> > +    return ENV.from_string(template, {
>
> I'm not sure if I understand what you mean by "variables that do not exist
> will be replaced with empty strings". Where is that happening?
>
> I am trying to implement the Rest sensor using the new templates. If a
> user provides {{ value_json.something_that_doesnt_exist }} or {{
> something_that_doesnt_exist }} as the template, the from_string method
> will raise an exception. I think it is acceptable to raise an exception in
> this cause because there is nothing else to do. We just want the exception
> to be of our own type, and not jinja2.exceptions.UndefinedError so that
> not every module has to import jinja2.
>
> I'm probably missing something.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/720/files#r47401827>.
>